bokeh/models/contour_renderer.py

```diff
@@ -0,0 +1,120 @@
+#-----------------------------------------------------------------------------
+# Copyright (c) 2012 - 2022, Anaconda, Inc., and Bokeh Contributors.
+# All rights reserved.
+#
+# The full license is in the file LICENSE.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+''' Models (mostly base classes) for the various kinds of renderer
+types that Bokeh supports.
+
+'''
+#-----------------------------------------------------------------------------
+# Boilerplate
+#-----------------------------------------------------------------------------
+from __future__ import annotations
+
+import logging # isort:skip
+log = logging.getLogger(__name__)
+
+#-----------------------------------------------------------------------------
+# Imports
+#-----------------------------------------------------------------------------
+
+# Bokeh imports
+from ..core.properties import Float, Instance, Seq
+from ..core.types import Unknown
+from .annotations import ContourColorBar
+from .glyphs import MultiLine, MultiPolygons
+from .renderers import DataRenderer, GlyphRenderer
+from .sources import ColumnDataSource
+from .tickers import FixedTicker
+
+#-----------------------------------------------------------------------------
+# Globals and constants
+#-----------------------------------------------------------------------------
+
+__all__ = (
+    'ContourRenderer',
+)
+
+#-----------------------------------------------------------------------------
+# General API
+#-----------------------------------------------------------------------------
+
+#-----------------------------------------------------------------------------
+# Dev API
+#-----------------------------------------------------------------------------
+
+_DEFAULT_CONTOUR_LINE_RENDERER = lambda: GlyphRenderer(
+    glyph=MultiLine(), data_source=ColumnDataSource(data=dict())
+)
+
+_DEFAULT_CONTOUR_FILL_RENDERER = lambda: GlyphRenderer(
+    glyph=MultiPolygons(), data_source=ColumnDataSource(data=dict())
+)
+
+class ContourRenderer(DataRenderer):
+    '''
+    Renderer for contour plots composed of filled polygons and/or lines.
+    Rather than create these manually it is usually better to use
+    ``figure.contour`` instead.
+    '''
+
+    # explicit __init__ to support Init signatures
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+    line_renderer = Instance(GlyphRenderer, default=_DEFAULT_CONTOUR_LINE_RENDERER, help="""
+    Glyph renderer used for contour lines.
+    """)
+
+    fill_renderer = Instance(GlyphRenderer, default=_DEFAULT_CONTOUR_FILL_RENDERER, help="""
+    Glyph renderer used for filled contour polygons.
+    """)
+
+    levels = Seq(Float, default=[], help="""
+    Levels at which the contours are calculated.
+    """)
+
+    def __setattr__(self, name: str, value: Unknown) -> None:
```

Comment:
Ah I was looking for an example and it was in the comments:
```
    new_contour_data = contour_data(x, y, z, levels)
    contour_renderer.data = new_contour_data
```
I think this dovetails with my suggestion to have `contour` data return a`dataclass`, then I think this could be accomplished with a property setter rather than a `__setattr__`?

Reply:
Done the `dataclass` and `set_data` approach.